File: Examples/Advanced/Archive-Configuration-Revisions/nodeJS/archive-configuration-revisions.mjs
#!/usr/bin/env node
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

export const STATE_FORMAT = "logicvein-netld-configuration-archive-state";
export const RUN_FORMAT = "logicvein-netld-configuration-archive-run";
const VERSION = 1;

export function loadEnv(file) {
  if (!fs.existsSync(file)) return;
  for (const raw of fs.readFileSync(file, "utf8").split(/\r?\n/)) {
    const match = raw.match(/^\s*(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)\s*=\s*(.*)\s*$/);
    if (!match || raw.trimStart().startsWith("#")) continue;
    let value = match[2];
    if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) value = value.slice(1, -1);
    process.env[match[1]] = value;
  }
}

const safeName = (value, fallback) => value.trim().replace(/[^A-Za-z0-9._-]+/g, "_").replace(/^[._]+|[._]+$/g, "") || fallback;
const pathHash = value => crypto.createHash("sha256").update(value).digest("hex").slice(0, 8);
const asPath = (base, value) => path.isAbsolute(value) ? value : path.join(base, value);

export function configFromEnv(base) {
  const required = ["NETLD_BASE_URL", "NETLD_API_KEY"];
  for (const name of required) if (!process.env[name]?.trim()) throw new Error(`${name} is required.`);
  const networks = [...new Set((process.env.NETLD_NETWORKS || "Default").split(",").map(v => v.trim()).filter(Boolean))].sort();
  const initialMode = (process.env.NETLD_INITIAL_MODE || "latest").trim().toLowerCase();
  if (!networks.length) throw new Error("NETLD_NETWORKS must contain at least one managed network.");
  if (!["latest", "all"].includes(initialMode)) throw new Error("NETLD_INITIAL_MODE must be either latest or all.");
  const inventoryPageSize = Number(process.env.NETLD_INVENTORY_PAGE_SIZE || 500);
  const historyPageSize = Number(process.env.NETLD_HISTORY_PAGE_SIZE || 500);
  if (!Number.isInteger(inventoryPageSize) || inventoryPageSize < 1 || !Number.isInteger(historyPageSize) || historyPageSize < 1) throw new Error("Page sizes must be positive integers.");
  return {
    baseUrl: process.env.NETLD_BASE_URL.trim().replace(/\/$/, ""), apiKey: process.env.NETLD_API_KEY.trim(), networks,
    archiveDir: asPath(base, process.env.NETLD_ARCHIVE_DIR || "configuration-archive"),
    statePath: asPath(base, process.env.NETLD_STATE_FILE || "configuration-archive-state.json"),
    runReportPath: asPath(base, process.env.NETLD_RUN_REPORT_FILE || "configuration-archive-run.json"),
    inventoryPageSize, historyPageSize, searchScheme: process.env.NETLD_SEARCH_SCHEME?.trim() || "ipAddress",
    searchQuery: process.env.NETLD_SEARCH_QUERY || "", initialMode,
  };
}

export class NetLDClient {
  constructor(config) { this.config = config; this.cookie = ""; }
  async request(method, body) {
    const headers = { Authorization: `Bearer ${this.config.apiKey}` };
    if (this.cookie) headers.Cookie = this.cookie;
    if (body) headers["Content-Type"] = "application/json";
    const response = await fetch(`${this.config.baseUrl}/rest`, { method, headers, body: body && JSON.stringify(body), redirect: "manual" });
    const cookie = response.headers.get("set-cookie"); if (cookie) this.cookie = cookie.split(";", 1)[0];
    if (!response.ok) throw new Error(`Request failed with HTTP ${response.status}.`);
    return response;
  }
  async login() { await this.request("GET"); }
  async call(method, params) {
    const response = await this.request("POST", { jsonrpc: "2.0", method, params, id: crypto.randomUUID() });
    const data = await response.json();
    if (data.error) throw new Error(`${method} failed: ${JSON.stringify(data.error)}`);
    return data.result;
  }
  searchInventory(config, offset) {
    const query = config.searchQuery.endsWith("\n") ? config.searchQuery : `${config.searchQuery}\n`;
    return this.call("Inventory.search", { network: config.networks, scheme: config.searchScheme, query,
      pageData: { offset, pageSize: config.inventoryPageSize }, sortColumn: "ipAddress", descending: false });
  }
  configurationHistory(device, offset, pageSize) {
    return this.call("Configuration.retrieveConfigHistory", { pageData: { offset, pageSize, total: 0, configHistoryItems: [] },
      networks: [device.network], scheme: "ipAddress", data: device.ipAddress, sortColumn: "session", descending: true });
  }
  retrieveRevision(item) {
    return this.call("Configuration.retrieveRevision", { network: item.managedNetwork, ipAddress: item.ipAddress,
      configPath: item.path, timestamp: item.lastChanged });
  }
}

async function collectPages(fetchPage, pageSize, listName, stopBefore = null) {
  const output = []; let offset = 0; let total = null;
  while (true) {
    const page = await fetchPage(offset); const values = page?.[listName] || [];
    for (const value of values) {
      if (stopBefore !== null && Number(value.lastChanged) < stopBefore) return output;
      output.push(value);
    }
    if (total === null && page?.total !== undefined && page.total !== null) total = Number(page.total);
    const actualPageSize = Number(page?.pageSize || pageSize);
    if ((total !== null && offset + values.length >= total) || (total === null && values.length < actualPageSize)) return output;
    if (!values.length) throw new Error(`${listName} returned an empty page before the reported total.`);
    offset += actualPageSize;
  }
}

export function selectCandidates(items, stateEntry, initialMode) {
  let selected;
  if (!stateEntry) {
    if (initialMode === "all") selected = items;
    else { const newest = new Map(); for (const item of items) if (!newest.has(item.path)) newest.set(item.path, item); selected = [...newest.values()]; }
  } else {
    const paths = new Set(stateEntry.pathsAtLastChanged || []), watermark = Number(stateEntry.lastChanged);
    selected = items.filter(i => Number(i.lastChanged) > watermark || (Number(i.lastChanged) === watermark && !paths.has(i.path)));
  }
  const unique = new Map(selected.map(i => [`${Number(i.lastChanged)}\0${i.path}`, i]));
  return [...unique.values()].sort((a, b) => Number(a.lastChanged) - Number(b.lastChanged) || a.path.localeCompare(b.path));
}

function loadState(file) {
  if (!fs.existsSync(file)) return { format: STATE_FORMAT, formatVersion: VERSION, devices: {} };
  const state = JSON.parse(fs.readFileSync(file, "utf8"));
  if (state.format !== STATE_FORMAT || state.formatVersion !== VERSION || !state.devices || Array.isArray(state.devices)) throw new Error("The archive state file has an unsupported format.");
  return state;
}
function atomicWrite(file, content) {
  fs.mkdirSync(path.dirname(file), { recursive: true }); const temporary = path.join(path.dirname(file), `.${path.basename(file)}.${crypto.randomUUID()}.tmp`);
  try { fs.writeFileSync(temporary, content, { mode: 0o600 }); fs.renameSync(temporary, file); } catch (error) { try { fs.unlinkSync(temporary); } catch {} throw error; }
}
const writeJson = (file, value) => atomicWrite(file, `${JSON.stringify(value, null, 2)}\n`);

function archiveOne(config, item, revision) {
  const encoded = revision.content || "";
  if (!/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(encoded)) throw new Error("Configuration revision content is not valid Base64.");
  const content = Buffer.from(encoded, "base64"), stem = `${Number(item.lastChanged)}_${safeName(item.path, "config")}_${pathHash(item.path)}`;
  const extension = String(revision.mimeType || item.mimeType || "").startsWith("text/") ? ".txt" : ".bin";
  const directory = path.join(config.archiveDir, safeName(item.managedNetwork, "network"), safeName(item.ipAddress, "device"));
  const contentPath = path.join(directory, stem + extension), metadataPath = path.join(directory, `${stem}.metadata.json`);
  atomicWrite(contentPath, content);
  const { content: ignored, ...revisionMetadata } = revision;
  writeJson(metadataPath, { network: item.managedNetwork, ipAddress: item.ipAddress, configPath: item.path,
    lastChanged: Number(item.lastChanged), history: item, revision: revisionMetadata, contentFile: path.basename(contentPath) });
  return { network: item.managedNetwork, ipAddress: item.ipAddress, configPath: item.path, lastChanged: Number(item.lastChanged),
    mimeType: revision.mimeType, size: content.length, contentFile: path.relative(config.archiveDir, contentPath), metadataFile: path.relative(config.archiveDir, metadataPath) };
}

export async function archiveConfigurationRevisions(client, config, generatedAt = new Date().toISOString().replace(/\.\d{3}Z$/, "Z")) {
  const state = loadState(config.statePath), archived = [], failures = [];
  const devices = await collectPages(offset => client.searchInventory(config, offset), config.inventoryPageSize, "devices");
  for (const device of devices) {
    const key = `${device.network}@${device.ipAddress}`, oldEntry = state.devices[key]; let candidates;
    try {
      const items = await collectPages(offset => client.configurationHistory(device, offset, config.historyPageSize), config.historyPageSize, "configHistoryItems", oldEntry ? Number(oldEntry.lastChanged) : null);
      candidates = selectCandidates(items, oldEntry, config.initialMode);
    } catch (error) { failures.push({ stage: "history", network: device.network, ipAddress: device.ipAddress, error: error.message }); continue; }
    let failed = false;
    for (const item of candidates) try { archived.push(archiveOne(config, item, await client.retrieveRevision(item))); }
    catch (error) { failed = true; failures.push({ stage: "revision", network: item.managedNetwork, ipAddress: item.ipAddress, configPath: item.path, lastChanged: item.lastChanged, error: error.message }); }
    if (candidates.length && !failed) {
      const newest = Math.max(...candidates.map(i => Number(i.lastChanged))); let paths = candidates.filter(i => Number(i.lastChanged) === newest).map(i => i.path);
      if (oldEntry && newest === Number(oldEntry.lastChanged)) paths.push(...(oldEntry.pathsAtLastChanged || []));
      state.devices[key] = { lastChanged: newest, pathsAtLastChanged: [...new Set(paths)].sort() };
    }
  }
  const report = { format: RUN_FORMAT, formatVersion: VERSION, generatedAt, initialMode: config.initialMode, deviceCount: devices.length,
    archivedCount: archived.length, failureCount: failures.length, archived, failures };
  writeJson(config.statePath, state); writeJson(config.runReportPath, report); return report;
}

async function main() {
  const here = path.dirname(fileURLToPath(import.meta.url)); const index = process.argv.indexOf("--env"); const envFile = index >= 0 ? process.argv[index + 1] : path.join(here, ".env");
  loadEnv(envFile); const config = configFromEnv(here); const client = new NetLDClient(config); await client.login(); const report = await archiveConfigurationRevisions(client, config);
  console.log(`Processed ${report.deviceCount} devices and archived ${report.archivedCount} revisions`); console.log(`Recorded ${report.failureCount} failures in ${config.runReportPath}`); process.exitCode = report.failureCount ? 2 : 0;
}
if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) main().catch(error => { console.error(`Error: ${error.message}`); process.exitCode = 1; });

File: Examples/Advanced/Archive-Configuration-Revisions/Python/archive_configuration_revisions.py
#!/usr/bin/env python3
"""Archive netLD/ThirdEye configuration revisions using documented JSON-RPC APIs."""

from __future__ import annotations

import argparse
import base64
import binascii
import hashlib
import json
import os
import re
import sys
import tempfile
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterator

import requests
from dotenv import load_dotenv


STATE_FORMAT = "logicvein-netld-configuration-archive-state"
RUN_FORMAT = "logicvein-netld-configuration-archive-run"
FORMAT_VERSION = 1


class ExampleError(RuntimeError):
    pass


def parse_networks(value: str) -> list[str]:
    networks = sorted({item.strip() for item in value.split(",") if item.strip()})
    if not networks:
        raise ExampleError("NETLD_NETWORKS must contain at least one managed network.")
    return networks


def safe_name(value: str, fallback: str) -> str:
    cleaned = re.sub(r"[^A-Za-z0-9._-]+", "_", value.strip()).strip("._")
    return cleaned or fallback


def path_hash(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()[:8]


def utc_timestamp() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


@dataclass(frozen=True)
class Config:
    base_url: str
    api_key: str
    networks: list[str]
    archive_dir: Path
    state_path: Path
    run_report_path: Path
    inventory_page_size: int
    history_page_size: int
    search_scheme: str
    search_query: str
    initial_mode: str

    @classmethod
    def from_env(cls, env_path: Path, output_base: Path) -> "Config":
        load_dotenv(env_path, override=True)
        base_url = os.getenv("NETLD_BASE_URL", "").strip()
        api_key = os.getenv("NETLD_API_KEY", "").strip()
        if not base_url or not api_key:
            raise ExampleError("Set NETLD_BASE_URL and NETLD_API_KEY in the environment file.")
        try:
            inventory_page_size = int(os.getenv("NETLD_INVENTORY_PAGE_SIZE", "500"))
            history_page_size = int(os.getenv("NETLD_HISTORY_PAGE_SIZE", "500"))
        except ValueError as exc:
            raise ExampleError("Page sizes must be positive integers.") from exc
        if inventory_page_size <= 0 or history_page_size <= 0:
            raise ExampleError("Page sizes must be positive integers.")
        initial_mode = os.getenv("NETLD_INITIAL_MODE", "latest").strip().lower()
        if initial_mode not in {"latest", "all"}:
            raise ExampleError("NETLD_INITIAL_MODE must be either latest or all.")

        def destination(name: str, default: str) -> Path:
            value = Path(os.getenv(name, default))
            return value if value.is_absolute() else output_base / value

        return cls(
            base_url.rstrip("/"), api_key,
            parse_networks(os.getenv("NETLD_NETWORKS", "Default")),
            destination("NETLD_ARCHIVE_DIR", "configuration-archive"),
            destination("NETLD_STATE_FILE", "configuration-archive-state.json"),
            destination("NETLD_RUN_REPORT_FILE", "configuration-archive-run.json"),
            inventory_page_size, history_page_size,
            os.getenv("NETLD_SEARCH_SCHEME", "ipAddress").strip() or "ipAddress",
            os.getenv("NETLD_SEARCH_QUERY", ""), initial_mode,
        )


class NetLDClient:
    def __init__(self, base_url: str, api_key: str, timeout: float = 30):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"Authorization": f"Bearer {api_key}"})

    def _request(self, method: str, **kwargs: Any):
        try:
            response = self.session.request(
                method, f"{self.base_url}/rest", timeout=self.timeout,
                allow_redirects=False, **kwargs,
            )
        except requests.RequestException as exc:
            raise ExampleError(f"Could not reach {self.base_url}.") from exc
        if 300 <= response.status_code < 400:
            raise ExampleError(f"Request redirected to {response.headers.get('Location', '')}.")
        try:
            response.raise_for_status()
        except requests.RequestException as exc:
            raise ExampleError(f"Request failed with HTTP {response.status_code}.") from exc
        return response

    def login(self) -> None:
        self._request("GET")

    def call(self, method: str, **parameters: Any) -> Any:
        response = self._request(
            "POST", headers={"Content-Type": "application/json"},
            json={"jsonrpc": "2.0", "method": method, "params": parameters, "id": str(uuid.uuid4())},
        )
        try:
            data = response.json()
        except ValueError as exc:
            raise ExampleError(f"{method} returned invalid JSON.") from exc
        if data.get("error"):
            raise ExampleError(f"{method} failed: {json.dumps(data['error'])}")
        return data.get("result")

    def search_inventory(self, config: Config, offset: int):
        query = config.search_query
        result = self.call(
            "Inventory.search", network=config.networks, scheme=config.search_scheme,
            query=query if query.endswith("\n") else f"{query}\n",
            pageData={"offset": offset, "pageSize": config.inventory_page_size},
            sortColumn="ipAddress", descending=False,
        )
        if not isinstance(result, dict):
            raise ExampleError("Inventory.search returned no page data.")
        return result

    def configuration_history(self, device: dict[str, Any], offset: int, page_size: int):
        result = self.call(
            "Configuration.retrieveConfigHistory",
            pageData={"offset": offset, "pageSize": page_size, "total": 0, "configHistoryItems": []},
            networks=[device["network"]], scheme="ipAddress", data=device["ipAddress"],
            sortColumn="session", descending=True,
        )
        if not isinstance(result, dict):
            raise ExampleError("Configuration.retrieveConfigHistory returned no page data.")
        return result

    def retrieve_revision(self, item: dict[str, Any]):
        result = self.call(
            "Configuration.retrieveRevision", network=item["managedNetwork"],
            ipAddress=item["ipAddress"], configPath=item["path"], timestamp=item["lastChanged"],
        )
        if not isinstance(result, dict):
            raise ExampleError("Configuration.retrieveRevision returned no revision.")
        return result


def inventory_devices(client: Any, config: Config) -> Iterator[dict[str, Any]]:
    offset = 0
    total: int | None = None
    while True:
        page = client.search_inventory(config, offset)
        devices = page.get("devices") or []
        for device in devices:
            yield device
        page_size = int(page.get("pageSize") or config.inventory_page_size)
        if total is None and page.get("total") is not None:
            total = int(page["total"])
        if total is not None and offset + len(devices) >= total:
            return
        if total is None and len(devices) < page_size:
            return
        if not devices:
            raise ExampleError("Inventory.search returned an empty page before the reported total.")
        offset += page_size


def history_items(
    client: Any, config: Config, device: dict[str, Any], stop_before: int | None = None,
) -> Iterator[dict[str, Any]]:
    offset = 0
    total: int | None = None
    while True:
        page = client.configuration_history(device, offset, config.history_page_size)
        items = page.get("configHistoryItems") or []
        for item in items:
            # Results are requested newest-first. Once an incremental scan passes
            # its watermark, no later page can contain a candidate.
            if stop_before is not None and int(item["lastChanged"]) < stop_before:
                return
            yield item
        page_size = int(page.get("pageSize") or config.history_page_size)
        if total is None and page.get("total") is not None:
            total = int(page["total"])
        if total is not None and offset + len(items) >= total:
            return
        if total is None and len(items) < page_size:
            return
        if not items:
            raise ExampleError("Configuration history returned an empty page before the reported total.")
        offset += page_size


def load_state(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {"format": STATE_FORMAT, "formatVersion": FORMAT_VERSION, "devices": {}}
    try:
        state = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError) as exc:
        raise ExampleError(f"Could not read archive state: {path}") from exc
    if state.get("format") != STATE_FORMAT or state.get("formatVersion") != FORMAT_VERSION:
        raise ExampleError("The archive state file has an unsupported format.")
    if not isinstance(state.get("devices"), dict):
        raise ExampleError("The archive state file has an invalid devices object.")
    return state


def write_atomic(path: Path, content: bytes) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary_path: Path | None = None
    try:
        with tempfile.NamedTemporaryFile(dir=path.parent, prefix=f".{path.name}.", suffix=".tmp", delete=False) as output:
            temporary_path = Path(output.name)
            output.write(content)
        os.replace(temporary_path, path)
    except Exception:
        if temporary_path is not None:
            temporary_path.unlink(missing_ok=True)
        raise


def write_json_atomic(path: Path, document: dict[str, Any]) -> None:
    write_atomic(path, (json.dumps(document, indent=2, sort_keys=True, ensure_ascii=False) + "\n").encode("utf-8"))


def select_candidates(
    items: list[dict[str, Any]], state_entry: dict[str, Any] | None, initial_mode: str,
) -> list[dict[str, Any]]:
    if state_entry is None:
        if initial_mode == "all":
            selected = items
        else:
            newest_by_path: dict[str, dict[str, Any]] = {}
            for item in items:
                newest_by_path.setdefault(item["path"], item)
            selected = list(newest_by_path.values())
    else:
        watermark = int(state_entry["lastChanged"])
        paths_at_watermark = set(state_entry.get("pathsAtLastChanged") or [])
        selected = [
            item for item in items
            if int(item["lastChanged"]) > watermark
            or (int(item["lastChanged"]) == watermark and item["path"] not in paths_at_watermark)
        ]
    unique = {(int(item["lastChanged"]), item["path"]): item for item in selected}
    return [unique[key] for key in sorted(unique)]


def archive_one_revision(config: Config, item: dict[str, Any], revision: dict[str, Any]) -> dict[str, Any]:
    try:
        content = base64.b64decode(revision.get("content") or "", validate=True)
    except (binascii.Error, ValueError) as exc:
        raise ExampleError("Configuration revision content is not valid Base64.") from exc
    network = item["managedNetwork"]
    ip_address = item["ipAddress"]
    config_path = item["path"]
    stem = f"{int(item['lastChanged'])}_{safe_name(config_path, 'config')}_{path_hash(config_path)}"
    extension = ".txt" if (revision.get("mimeType") or item.get("mimeType") or "").startswith("text/") else ".bin"
    directory = config.archive_dir / safe_name(network, "network") / safe_name(ip_address, "device")
    content_path = directory / f"{stem}{extension}"
    metadata_path = directory / f"{stem}.metadata.json"
    write_atomic(content_path, content)
    metadata = {
        "network": network, "ipAddress": ip_address, "configPath": config_path,
        "lastChanged": int(item["lastChanged"]),
        "history": item,
        "revision": {key: value for key, value in revision.items() if key != "content"},
        "contentFile": content_path.name,
    }
    write_json_atomic(metadata_path, metadata)
    return {
        "network": network, "ipAddress": ip_address, "configPath": config_path,
        "lastChanged": int(item["lastChanged"]), "mimeType": revision.get("mimeType"),
        "size": len(content),
        "contentFile": str(content_path.relative_to(config.archive_dir)),
        "metadataFile": str(metadata_path.relative_to(config.archive_dir)),
    }


def archive_configuration_revisions(
    client: Any, config: Config, generated_at: str | None = None,
) -> dict[str, Any]:
    state = load_state(config.state_path)
    archived: list[dict[str, Any]] = []
    failures: list[dict[str, Any]] = []
    device_count = 0
    for device in inventory_devices(client, config):
        device_count += 1
        key = f"{device['network']}@{device['ipAddress']}"
        old_entry = state["devices"].get(key)
        try:
            stop_before = int(old_entry["lastChanged"]) if old_entry else None
            all_items = list(history_items(client, config, device, stop_before))
            candidates = select_candidates(all_items, old_entry, config.initial_mode)
        except Exception as exc:
            failures.append({
                "stage": "history", "network": device.get("network"),
                "ipAddress": device.get("ipAddress"), "error": str(exc),
            })
            continue
        device_archived = []
        device_failed = False
        for item in candidates:
            try:
                revision = client.retrieve_revision(item)
                record = archive_one_revision(config, item, revision)
                archived.append(record)
                device_archived.append(record)
            except Exception as exc:
                device_failed = True
                failures.append({
                    "stage": "revision", "network": item.get("managedNetwork"),
                    "ipAddress": item.get("ipAddress"), "configPath": item.get("path"),
                    "lastChanged": item.get("lastChanged"), "error": str(exc),
                })
        if candidates and not device_failed:
            newest = max(int(item["lastChanged"]) for item in candidates)
            paths = sorted({item["path"] for item in candidates if int(item["lastChanged"]) == newest})
            if old_entry and newest == int(old_entry["lastChanged"]):
                paths = sorted(set(paths) | set(old_entry.get("pathsAtLastChanged") or []))
            state["devices"][key] = {"lastChanged": newest, "pathsAtLastChanged": paths}
    report = {
        "format": RUN_FORMAT, "formatVersion": FORMAT_VERSION,
        "generatedAt": generated_at or utc_timestamp(), "initialMode": config.initial_mode,
        "deviceCount": device_count, "archivedCount": len(archived),
        "failureCount": len(failures), "archived": archived, "failures": failures,
    }
    write_json_atomic(config.state_path, state)
    write_json_atomic(config.run_report_path, report)
    return report


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--env", type=Path, default=Path(__file__).with_name(".env"))
    args = parser.parse_args()
    try:
        config = Config.from_env(args.env, Path(__file__).resolve().parent)
        client = NetLDClient(config.base_url, config.api_key)
        client.login()
        report = archive_configuration_revisions(client, config)
        print(f"Processed {report['deviceCount']} devices and archived {report['archivedCount']} revisions")
        print(f"Recorded {report['failureCount']} failures in {config.run_report_path}")
        return 2 if report["failureCount"] else 0
    except (ExampleError, OSError, ValueError, KeyError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
